File: java/dsframework/src/main/java/dsf/route/HotStandby.java
package dsf.route;

import dsf.register.Service;
import dsf.register.ServiceInstance;
import java.util.List;
import java.util.TreeSet;

/**
 * 热备的路由策略，按服务器地址排序，选择第一个在线的服务
 * @author arksea
 */
public class HotStandby extends AbstractRouteStrategy {
    
    public HotStandby(Service s) {
        super(s);
    }

    @Override
    public ServiceInstance getInstance() {
        List<ServiceInstance> list = service.getInstanceList();
        TreeSet<ServiceInstance> sorted = new TreeSet();
        sorted.addAll(list);
        int onlineCount = 0;
        for (ServiceInstance s : sorted) {
            String onlineStr = (String) s.getState("online");
            boolean online = (onlineStr != null && onlineStr.equals("true"));
            if (online) {
                ++onlineCount;
                if (s.tryAcquire()) {
                    return s;
                }
            }
        }
        if (onlineCount == 0) throw new NoUseableServiceException(service.define.name);
        throw new RateLimitedException(service.define.name);
    }
}

File: csharp/dsf/dsf/src/route/HotStandby.cs
using System.Collections.Generic;

namespace com.baidu.dsf.route
{

	using Service = com.baidu.dsf.register.Service;
	using ServiceInstance = com.baidu.dsf.register.ServiceInstance;

	/// <summary>
	/// 热备的路由策略，按服务器地址排序，选择第一个在线的服务
	/// @author arksea
	/// </summary>
	public class HotStandby : AbstractRouteStrategy
	{

		public HotStandby(Service s) : base(s)
		{
		}

		public override ServiceInstance Instance
		{
			get
			{
				java.util.List list = service.instanceList;
                SortedSet<ServiceInstance> sorted = new SortedSet<ServiceInstance>();
                java.util.Iterator it = list.iterator();
                while(it.hasNext())
                {
                    ServiceInstance s = (ServiceInstance)it.next();
                    sorted.Add(s);
                }
				int onlineCount = 0;
				foreach (ServiceInstance s in sorted)
				{
					string onlineStr = (string) s.getState("online");
					bool online = (onlineStr != null && onlineStr.Equals("true"));
					if (online)
					{
						++onlineCount;
						if (s.tryAcquire())
						{
							return s;
						}
					}
				}
				if (onlineCount == 0)
				{
					throw new NoUseableServiceException(service.define.name);
				}
				throw new RateLimitedException(service.define.name);
			}
		}
	}

}